zebra-state/src/service/finalized_state/zebra_db/block.rs

```diff
@@ -517,18 +517,26 @@ impl ZebraDb {
             .collect();
 
         // Get the current address balances, before the transactions in this block
-        let address_balances: HashMap<transparent::Address, AddressBalanceLocationChange> =
+
+        fn read_addr_locs<T, F: Fn(&transparent::Address) -> Option<T>>(
+            changed_addresses: HashSet<transparent::Address>,
+            f: F,
+        ) -> HashMap<transparent::Address, T> {
             changed_addresses
                 .into_iter()
-                .filter_map(|address| {
-                    // # Correctness
-                    //
-                    // Address balances are updated with the `fetch_add_balance_and_received` merge operator, so
-                    // the values must represent the changes to the balance, not the final balance.
-                    let addr_loc = self.address_balance_location(&address)?.into_new_change();
-                    Some((address.clone(), addr_loc))
-                })
-                .collect();
+                .filter_map(|address| Some((address.clone(), f(&address)?)))
+                .collect()
+        }
+
+        let address_balances: AddressBalanceLocationUpdates = if self.finished_format_upgrades() {
+            AddressBalanceLocationUpdates::Insert(read_addr_locs(changed_addresses, |addr| {
+                self.address_balance_location(addr)
+            }))
+        } else {
+            AddressBalanceLocationUpdates::Merge(read_addr_locs(changed_addresses, |addr| {
+                Some(self.address_balance_location(addr)?.into_new_change())
+            }))
+        };
```

Comment:
Could you add docs describing why we use `Insert` when `finished_format_upgrades` returns `true` and `Merge` otherwise?